File: benchmarks/references/ground-truth/Python/tc_017_shift_table.py
#!/usr/bin/env python3
"""TC-017: Laboratory Shift Table
Baseline vs worst post-baseline lab category shifts (LOW/NORMAL/HIGH).
Output: JSON with shift counts per cell, overall percentages.

Dependencies: numpy, pandas
"""

import argparse
import json
import sys
import numpy as np
import pandas as pd


def generate_labs(n: int, seed: int) -> pd.DataFrame:
    rng = np.random.default_rng(seed)
    n_exp = n // 2
    n_ctl = n - n_exp
    trt = np.array([1] * n_exp + [0] * n_ctl)
    rng.shuffle(trt)

    # Baseline hemoglobin (g/L)
    baseline_hgb = rng.normal(140, 18, size=n)

    # 3 post-baseline visits; worst = min for hemoglobin
    post_hgb = np.empty(n)
    for i in range(n):
        effect = -5 if trt[i] == 1 else 0
        vals = rng.normal(baseline_hgb[i] + effect, 10, size=3)
        post_hgb[i] = vals.min()

    # Categorize
    def categorize(x_arr):
        cats = np.empty(len(x_arr), dtype=object)
        cats[x_arr < 120] = "LOW"
        cats[(x_arr >= 120) & (x_arr <= 160)] = "NORMAL"
        cats[x_arr > 160] = "HIGH"
        return cats

    bl_cat = categorize(baseline_hgb)
    post_cat = categorize(post_hgb)

    return pd.DataFrame({
        "USUBJID": [f"SUB{i:04d}" for i in range(1, n + 1)],
        "TRT01PN": trt.astype(int),
        "TRT01A": np.where(trt == 1, "Experimental", "Control"),
        "LBTEST": "Hemoglobin",
        "BL_VAL": np.round(baseline_hgb, 2),
        "BL_CAT": bl_cat,
        "POST_VAL": np.round(post_hgb, 2),
        "POST_CAT": post_cat,
    })


def categorize_val(x):
    if x < 120:
        return "LOW"
    elif x > 160:
        return "HIGH"
    return "NORMAL"


def build_shift(d: pd.DataFrame) -> dict:
    cats = ["LOW", "NORMAL", "HIGH"]
    counts = {bl: {post: 0 for post in cats} for bl in cats}

    for _, row in d.iterrows():
        bl = row["BL_CAT"]
        post = row["POST_CAT"]
        if bl in cats and post in cats:
            counts[bl][post] += 1

    n_total = sum(counts[bl][post] for bl in cats for post in cats)
    percentages = {}
    for bl in cats:
        percentages[bl] = {}
        for post in cats:
            percentages[bl][post] = round(counts[bl][post] / n_total * 100, 2) if n_total > 0 else 0.0

    return {
        "counts": counts,
        "percentages": percentages,
        "n_total": int(n_total),
        "n_stable_normal": int(counts["NORMAL"]["NORMAL"]),
        "n_low_to_normal": int(counts["LOW"]["NORMAL"]),
        "n_normal_to_low": int(counts["NORMAL"]["LOW"]),
        "n_normal_to_high": int(counts["NORMAL"]["HIGH"]),
        "n_high_to_normal": int(counts["HIGH"]["NORMAL"]),
    }


def main():
    parser = argparse.ArgumentParser(description="TC-017: Lab Shift Table")
    parser.add_argument("--seed", type=int, default=42)
    parser.add_argument("--n", type=int, default=200)
    parser.add_argument("--output", type=str, default="")
    parser.add_argument("--data", type=str, default="")
    args = parser.parse_args()

    print(f"[{pd.Timestamp.now()}] TC-017 Lab Shift Table | seed={args.seed} n={args.n}")

    if args.data:
        df = pd.read_csv(args.data)
    else:
        df = generate_labs(args.n, args.seed)

    exp_df = df[df["TRT01PN"] == 1]
    ctl_df = df[df["TRT01PN"] == 0]

    result = {
        "tc_id": "TC-017",
        "tc_name": "Laboratory Shift Table",
        "metadata": {
            "n_total": int(len(df)),
            "n_experimental": int(len(exp_df)),
            "n_control": int(len(ctl_df)),
            "population": "SAFETY",
            "lab_test": "Hemoglobin",
            "lab_unit": "g/L",
            "categories": ["LOW", "NORMAL", "HIGH"],
            "thresholds": {"low": 120, "high": 160},
            "n_post_baseline_visits": 3,
            "python_version": sys.version.split()[0],
        },
        "shift_overall": build_shift(df),
        "shift_experimental": build_shift(exp_df),
        "shift_control": build_shift(ctl_df),
    }

    js = json.dumps(result, indent=2)
    if args.output:
        with open(args.output, "w") as f:
            f.write(js)
        print(f"Written: {args.output}")
    else:
        print(js)


if __name__ == "__main__":
    main()

File: benchmarks/references/ground-truth/R/tc-017-shift-table.R
#!/usr/bin/env Rscript
# TC-017: Laboratory Shift Table
# Baseline vs post-baseline lab category shifts (LOW/NORMAL/HIGH)
# Output: JSON with shift counts per cell, overall percentages

args <- commandArgs(trailingOnly = TRUE)
seed <- 42; n <- 200; output <- ""; data_csv <- ""
i <- 1
while (i <= length(args)) {
  if (args[i] == "--seed") { seed <- as.integer(args[i+1]); i <- i + 2 }
  else if (args[i] == "--n") { n <- as.integer(args[i+1]); i <- i + 2 }
  else if (args[i] == "--output") { output <- args[i+1]; i <- i + 2 }
  else if (args[i] == "--data") { data_csv <- args[i+1]; i <- i + 2 }
  else { i <- i + 1 }
}

cat(sprintf("[%s] TC-017 Lab Shift Table | seed=%d n=%d\n", Sys.time(), seed, n))

# --- Data generation ---
generate_labs <- function(n, seed) {
  set.seed(seed)
  # Hemoglobin (g/L): normal range ~120-160 (F), 130-180 (M)
  # Low <120, Normal 120-160, High >160
  n_subj <- n
  n_visits <- 4  # baseline + 3 post-baseline

  subjects <- sprintf("SUB%04d", seq_len(n_subj))
  trt <- rep(c(1, 0), each = n_subj %/% 2)
  trt <- sample(trt)

  # Baseline lab values
  baseline_hgb <- rnorm(n_subj, mean = 140, sd = 18)

  # Worst post-baseline (most extreme deviation from normal)
  post_hgb <- mapply(function(bl, t) {
    # Treatment effect: experimental arm tends to decrease
    effect <- ifelse(t == 1, -5, 0)
    vals <- rnorm(3, mean = bl + effect, sd = 10)  # 3 post-baseline visits
    min(vals)  # worst = lowest for hemoglobin
  }, baseline_hgb, trt)

  # Categorize
  categorize <- function(x) {
    ifelse(x < 120, "LOW", ifelse(x > 160, "HIGH", "NORMAL"))
  }

  bl_cat <- categorize(baseline_hgb)
  post_cat <- categorize(post_hgb)

  data.frame(
    USUBJID = subjects,
    TRT01PN = trt,
    TRT01A = ifelse(trt == 1, "Experimental", "Control"),
    LBTEST = "Hemoglobin",
    BL_VAL = round(baseline_hgb, 2),
    BL_CAT = bl_cat,
    POST_VAL = round(post_hgb, 2),
    POST_CAT = post_cat,
    stringsAsFactors = FALSE
  )
}

if (data_csv != "") {
  df <- read.csv(data_csv, stringsAsFactors = FALSE)
} else {
  df <- generate_labs(n, seed)
}

library(jsonlite)

# --- Build shift table ---
build_shift <- function(d) {
  cats <- c("LOW", "NORMAL", "HIGH")
  # Initialize 3x3 matrix of zeros
  mat <- matrix(0L, nrow = 3, ncol = 3,
    dimnames = list(Baseline = cats, PostBaseline = cats))

  for (i in seq_len(nrow(d))) {
    bl <- d$BL_CAT[i]; post <- d$POST_CAT[i]
    if (bl %in% cats && post %in% cats) {
      mat[bl, post] <- mat[bl, post] + 1L
    }
  }

  # Convert to list format
  rows <- lapply(cats, function(bl) {
    setNames(as.list(as.integer(mat[bl, ])), cats)
  })
  names(rows) <- cats

  n_total <- sum(mat)
  # Overall percentages
  pct <- round(mat / n_total * 100, 2)

  list(
    counts = rows,
    percentages = lapply(cats, function(bl) {
      setNames(as.list(as.numeric(pct[bl, ])), cats)
    }),
    n_total = as.integer(n_total),
    # Shift summary
    n_stable_normal = as.integer(mat["NORMAL", "NORMAL"]),
    n_low_to_normal = as.integer(mat["LOW", "NORMAL"]),
    n_normal_to_low = as.integer(mat["NORMAL", "LOW"]),
    n_normal_to_high = as.integer(mat["NORMAL", "HIGH"]),
    n_high_to_normal = as.integer(mat["HIGH", "NORMAL"])
  )
}

exp_shift <- build_shift(df[df$TRT01PN == 1, ])
ctl_shift <- build_shift(df[df$TRT01PN == 0, ])
overall_shift <- build_shift(df)

result <- list(
  tc_id = "TC-017",
  tc_name = "Laboratory Shift Table",
  metadata = list(
    n_total = nrow(df),
    n_experimental = sum(df$TRT01PN == 1),
    n_control = sum(df$TRT01PN == 0),
    population = "SAFETY",
    lab_test = "Hemoglobin",
    lab_unit = "g/L",
    categories = c("LOW", "NORMAL", "HIGH"),
    thresholds = list(low = 120, high = 160),
    n_post_baseline_visits = 3,
    r_version = R.version.string
  ),
  shift_overall = overall_shift,
  shift_experimental = exp_shift,
  shift_control = ctl_shift
)

# Integer metadata
result$metadata$n_total <- as.integer(result$metadata$n_total)
result$metadata$n_experimental <- as.integer(result$metadata$n_experimental)
result$metadata$n_control <- as.integer(result$metadata$n_control)

json <- toJSON(result, auto_unbox = TRUE, pretty = TRUE)

if (output != "") {
  writeLines(json, output)
  cat(sprintf("Written: %s\n", output))
} else {
  cat(json, "\n")
}
